### PR TITLE
feat: Modal footer support custom render function

### LINKED ISSUE #44297: Modal：是否考虑让 footer 支持传入 render 函数，或者支持 extraFooter
### What problem does this feature solve?

有挺多的时候可能只是想在 确认/取消 按钮之外多加一个额外的操作按钮，但是按目前的配置，必须得把整个 footer 重写，并且重写的过程也要自己实现一遍 确认/取消 按钮，包括自带的 确认/取消 按钮会触发的一些事件（比如 onOk/onCancel）等也要自己手动触发，有点麻烦。
如果支持 extraFooter 甚至让 footer 支持传入 render 函数（render 函数参数包括自带的 确认/取消 按钮），是不是可以更灵活和方便。

### What does the proposed API look like?

支持 extraFooter 属性或者让 footer 属性支持传入 render 函数（render 函数参数包括自带的 确认/取消 按钮）

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close #44297

### 💡 Background and solution
image
image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Modal footer support custom render function       |
| 🇨🇳 Chinese |   模态框页脚支持自定义函数渲染        |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 641db8d</samp>

This pull request adds a new feature to the `Modal` component and its methods, allowing users to customize the footer rendering function with the `footer` prop. It also refactors the footer buttons into separate components and uses context to pass the props. It updates the documentation, the type definitions, and the test cases accordingly.

### <samp>🤖 Generated by Copilot at 641db8d</samp>

*  Add custom footer rendering function feature to `Modal` and `Modal.confirm` methods ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d3734257bd9f28b52a331a7cc2fb5a3efd91c2fc71be29356d2fb171768ba9a3L128-R175), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cL97-R101), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L25-R28), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L104-R107), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5L56-R57), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5L106-R107), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3L57-R58), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3L107-R108), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-9986bf447ebc48c235cde6505736c4ce7affbb8eb8a06e94e5867cbd7fdb5626R1-R7))
  * Use `ConfirmCancelBtn` and `ConfirmOkBtn` components to replace `ActionButton` in `ConfirmContent` function ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d3734257bd9f28b52a331a7cc2fb5a3efd91c2fc71be29356d2fb171768ba9a3L103-R152), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-6691f460299001edf2b16e7eb0c931e8b83b62a914b008ea30137d165d4a6e0cR1-R53), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-71736387db0730fabaa7b99e3ccc9b8b27a66eaf90fa37cb49bc0dfe13e01db6R1-R51))
  * Use `NormalCancelBtn` and `NormalOkBtn` components to replace `Button` in `Footer` function ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-28fe880db90aca35d37607b3087393a1815128428f281a9a8394d5dc34475371L45-R99), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-4412ef1c364cdc5292e369c306bdc533f0a9ea2fed460bc17428c029c4260117R1-R25), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-bb2cdcb0b65a2807dff0587a2fbc3492482676baa9a6c4fcfceea49bde289f88R1-R33))
  * Use context and context providers to pass props to footer buttons ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d3734257bd9f28b52a331a7cc2fb5a3efd91c2fc71be29356d2fb171768ba9a3L15-R20), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-4ef6f5e7931e5a0f21fdbe8a76438b465de983cf564272b4d19a00fe66d73a08R1-R20), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-28fe880db90aca35d37607b3087393a1815128428f281a9a8394d5dc34475371L3-R11))
  * Support `footer` prop as a function that receives origin node and extra components as arguments ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d3734257bd9f28b52a331a7cc2fb5a3efd91c2fc71be29356d2fb171768ba9a3L128-R175), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cL97-R101), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L25-R28), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-ac9fa8e4784b40c5c2a26fe86dc50bbd39572b8ecf42b73611745ec817d883a1L104-R107), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5L56-R57), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5L106-R107), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3L57-R58), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3L107-R108), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-9986bf447ebc48c235cde6505736c4ce7affbb8eb8a06e94e5867cbd7fdb5626R1-R7), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d33a63c2b07982f75c8cec9567d82bcb1d2fa244617a37252109343a4ded9d6fR1-R65))
  * Add examples and documentation for the feature in both Chinese and English ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-9986bf447ebc48c235cde6505736c4ce7affbb8eb8a06e94e5867cbd7fdb5626R1-R7), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d33a63c2b07982f75c8cec9567d82bcb1d2fa244617a37252109343a4ded9d6fR1-R65), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5R25), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3R26))
* Add test cases to check custom footer rendering function feature in `Modal` and `Modal.confirm` methods ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-5b32c615603a02b35bc7b0b1651031eb23c5af3d939495dc127088d2efcd58b0R136-R151), [link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-c211b371c72d68e91c25e16529bf8ded303122c95060910f58afdfc61a942a2bR848-R864))
* Remove unused import of `ActionButton` from `ConfirmDialog` ([link](https://github.com/ant-design/ant-design/pull/44318/files?diff=unified&w=0#diff-d3734257bd9f28b52a331a7cc2fb5a3efd91c2fc71be29356d2fb171768ba9a3L7))

### CODE REVIEW

File: components/modal/index.zh-CN.md [hunk 1]
```diff
@@ -54,7 +55,7 @@ demo:
 | confirmLoading | 确定按钮 loading | boolean | false |  |
 | destroyOnClose | 关闭时销毁 Modal 里的子元素 | boolean | false |  |
 | focusTriggerAfterClose | 对话框关闭后是否需要聚焦触发元素 | boolean | true | 4.9.0 |
-| footer | 底部内容，当不需要默认底部按钮时，可以设为 `footer={null}` | ReactNode | (确定取消按钮) |  |
+| footer | 底部内容，当不需要默认底部按钮时，可以设为 `footer={null}` | (originNode: React.ReactNode, extra: { ConfirmBtn: FC; CancelBtn: FC }) => React.ReactNode \| React.ReactNode | (确定取消按钮) | 5.9.0 |
```
Comment (afc163) [inline components/modal/index.zh-CN.md:58]: 感觉需要有个 TS 定义放在 Popover 的交互比较好，这里面越写越长了。

**Review by kiner-tang — APPROVED** [review]

File: components/modal/index.zh-CN.md (hunk 1 above)
Comment (RedJue) [reply components/modal/index.zh-CN.md:58]: 可以，我加一个类型吧，然后链接过去

**Comment by afc163** [comment]
体积怎么多了这么多。

**Comment by RedJue** [comment]
> 体积怎么多了这么多。

代码层面感觉加的不多，是第三方依赖更新？

**Comment by afc163** [comment]
不是的，这个体积对比就只对比代码本身的变化。看上去拆了很多组件出来。

**Comment by RedJue** [comment]
> 不是的，这个体积对比就只对比代码本身的变化。看上去拆了很多新组件文件出来，看下是否有必要。

因为这个 `Api`  需要把原来写在一起的 `Button` 抽离成单个组件，好像也没法避免？

File: components/modal/ConfirmDialog.tsx [hunk 2]
```diff
@@ -100,20 +104,50 @@ export function ConfirmContent(
...
+    autoFocusButton,
+    cancelButtonProps,
+    cancelTextLocale,
+    isSilent,
+    mergedOkCancel,
+    rootPrefixCls,
+    close,
+    onCancel,
+    onConfirm,
+  };
```
Comment (afc163) [inline components/modal/ConfirmDialog.tsx:134]: 这两排属性声明了很多次，蛮占体积，是否是有必要的？

File: components/modal/components/ConfirmCancelBtn.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  const {
+    autoFocusButton,
+    cancelButtonProps,
+    cancelTextLocale,
+    isSilent,
+    mergedOkCancel,
+    rootPrefixCls,
+    close,
+    onCancel,
+    onConfirm,
```
Comment (afc163) [inline components/modal/components/ConfirmCancelBtn.tsx:32]: 这里也有一排

File: components/modal/ConfirmDialog.tsx (hunk 2 above)
Comment (RedJue) [reply components/modal/ConfirmDialog.tsx:134]: 我尝试去掉吧，再优化一下

File: components/modal/Modal.tsx [hunk 4]
```diff
@@ -94,7 +94,11 @@ const Modal: React.FC<ModalProps> = (props) => {
   }
 
   const dialogFooter =
-    footer === undefined ? <Footer {...props} onOk={handleOk} onCancel={handleCancel} /> : footer;
+    footer === undefined || typeof footer === 'function' ? (
+      <Footer {...props} onOk={handleOk} onCancel={handleCancel} />
+    ) : (
+      footer
+    );
```
Comment (Wxh16144) [inline components/modal/Modal.tsx:101]: 同上， 可以考虑抽象为一个函数来判断， 三元运算有点多了

File: components/modal/ConfirmDialog.tsx [hunk 5]
```diff
@@ -125,24 +140,20 @@ export function ConfirmContent(
...
+            <div className={`${confirmPrefixCls}-btns`}>
+              {footer === undefined
+                ? footerOriginNode
+                : footer?.(footerOriginNode, {
+                    ConfirmBtn,
+                    CancelBtn,
+                  })}
+            </div>
+          </ConfirmCancelBtnContextProvider>
+        </ConfirmOkBtnProvider>
```
Comment (Wxh16144) [inline components/modal/ConfirmDialog.tsx:156]: 这段代码感觉不是那么好阅读

File: components/modal/index.en-US.md [hunk 6]
```diff
@@ -180,6 +181,14 @@ return <div>{contextHolder}</div>;
...
 ```
 
+## footerRenderParams
+
+<!-- prettier-ignore -->
+| Property | Description | Type | Default |
+| --- | --- | --- | --- |
+| originNode | default node | React.ReactNode                   | -      |
+| extra      | extended options | { ConfirmBtn: FC; CancelBtn: FC } | -      |
+
```
Comment (Wxh16144) [inline components/modal/index.en-US.md:191]: 感觉这里可以直接用  tsx + interface , 比如：（伪代码）

```tsx
type FooterBuiltinProps = {
  ConfirmBtn: React.ReactNode;
  CancelBtn: React.ReactNode;
}
type RenderFooter = (originNode: FooterProps, extra: FooterBuiltinProps) => React.ReactNode;

export type FooterProps = React.ReactNode | RenderFooter;
```

File: components/modal/index.zh-CN.md [hunk 7]
```diff
@@ -181,6 +182,14 @@ return <div>{contextHolder}</div>;
...
 ```
 
+## footerRenderParams
+
+<!-- prettier-ignore -->
+| 参数 | 说明 | 类型 | 默认值 |
+| --- | --- | --- | --- |
+| originNode | 默认节点 | React.ReactNode                   | -      |
+| extra      | 扩展选项 | { ConfirmBtn: FC; CancelBtn: FC } | -      |
+
```
Comment (Wxh16144) [inline components/modal/index.zh-CN.md:192]: 同上👆🏻

File: components/modal/index.en-US.md (hunk 6 above)
Comment (RedJue) [reply components/modal/index.en-US.md:191]: 文档里好像很难显示ts定义

File: components/modal/shared.tsx [hunk 8]
```diff
@@ -42,23 +45,58 @@ export const Footer: React.FC<
...
-        loading={confirmLoading}
-        onClick={onOk}
-        {...okButtonProps}
-      >
-        {okText || locale?.okText}
-      </Button>
+      <NormalOkBtnContextProvider value={confirmBtnCtxValueMemo}>
+        <NormalCancelBtnContextProvider value={cancelBtnCtxValueMemo}>
+          {typeof footer === 'function'
+            ? footer?.(footerOriginNode, {
```
Comment (Wxh16144) [inline components/modal/shared.tsx:93]: 前面已经 typeof 了这里应该不需要判断了
```suggestion
            ? footer(footerOriginNode, {
```

File: components/modal/shared.tsx [hunk 9]
```diff
@@ -42,23 +45,58 @@ export const Footer: React.FC<
...
+
+  const cancelBtnCtxValue: NormalCancelBtnProps = {
+    cancelButtonProps,
+    cancelTextLocale,
+    onCancel,
+  };
+
+  const confirmBtnCtxValueMemo = React.useMemo(
+    () => confirmBtnCtxValue,
+    [...Object.values(confirmBtnCtxValue)],
```
Comment (Wxh16144) [inline components/modal/shared.tsx:74]: 这样写行不行呢？（其实我也不太熟悉 23333~）
```suggestion
    Object.values(confirmBtnCtxValue),
```

File: components/modal/shared.tsx [hunk 10]
```diff
@@ -42,23 +45,58 @@ export const Footer: React.FC<
...
+  );
+  const cancelBtnCtxValueMemo = React.useMemo(
+    () => cancelBtnCtxValue,
+    [...Object.values(cancelBtnCtxValue)],
```
Comment (Wxh16144) [inline components/modal/shared.tsx:78]:
```suggestion
    Object.values(cancelBtnCtxValue),
```

File: components/modal/index.en-US.md (hunk 6 above)
Comment (Wxh16144) [reply components/modal/index.en-US.md:191]: 那就先保持你的写法~

**Review by Wxh16144 — APPROVED** [review]

File: components/modal/shared.tsx (hunk 9 above)
Comment (RedJue) [reply components/modal/shared.tsx:74]: 感觉可以，看起来不需要解构

Comment (Wxh16144) [reply components/modal/shared.tsx:74]: 你先本地试试， confirmBtnCtxValue如果是 undefined values 是不是会报错

File: components/modal/shared.tsx [hunk 11]
```diff
@@ -42,23 +45,58 @@ export const Footer: React.FC<
...
+  );
+
+  const footerOriginNode = (
+    <>
+      <NormalCancelBtn />
+      <NormalOkBtn />
+    </>
+  );
```
[the remark this answers is not in the document]
Comment (RedJue) [reply components/modal/shared.tsx:86]: 大体逻辑差不多，正常的 `Modal` 和 `Modal.confirm` 两边按钮逻辑不太一样，所以又拆了两个按钮出来

File: components/modal/shared.tsx (hunk 9 above)
Comment (RedJue) [reply components/modal/shared.tsx:74]: 这是自己定义的，不会是 undefined

File: components/modal/context.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import type { NormalOkBtnProps } from './components/NormalOkBtn';
+import type { ConfirmCancelBtnProps } from './components/ConfirmCancelBtn';
+import type { ConfirmOkBtnProps } from './components/ConfirmOkBtn';
+
+export const NormalCancelBtnContext = React.createContext<NormalCancelBtnProps>(
+  {} as NormalCancelBtnProps,
+);
+export const NormalOkBtnContext = React.createContext<NormalOkBtnProps>({} as NormalOkBtnProps);
+export const ConfirmCancelBtnContext = React.createContext<ConfirmCancelBtnProps>(
+  {} as ConfirmCancelBtnProps,
```
Comment (MadCcc) [inline components/modal/context.ts:13]: 这里能不能只用一个 context，本身都是 Modal 的内部属性

Comment (RedJue) [reply components/modal/context.ts:13]: 应该可以收敛成一个，我一会优化一下

File: components/modal/components/NormalOkBtn.tsx [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  extends Pick<ModalProps, 'confirmLoading' | 'okType' | 'okButtonProps' | 'onOk'> {
+  okTextLocale?:
+    | string
+    | number
+    | true
+    | React.ReactElement<any, string | React.JSXElementConstructor<any>>
+    | Iterable<React.ReactNode>;
+}
+
+const NormalOkBtn: FC = () => {
```
Comment (MadCcc) [inline components/modal/components/NormalOkBtn.tsx:19]: 感觉 DefaultOkBtn 好点

File: components/modal/shared.tsx [hunk 14]
```diff
@@ -42,23 +45,46 @@ export const Footer: React.FC<
     onCancel,
     okButtonProps,
     cancelButtonProps,
+    footer,
```
Comment (MadCcc) [inline components/modal/shared.tsx:48]: 感觉不应该把 footer 放到 Footer 组件里处理，逻辑应该和 node 一样是整体替换掉的

Comment (MadCcc) [reply components/modal/shared.tsx:48]: 逻辑应该保持一致，要么都在里面处理，要么都在外面处理

Comment (RedJue) [reply components/modal/shared.tsx:48]: 那我都收到里面去统一处理

**Comment by afc163** [comment]
多了更多了，加这个功能多了 400B ..

**Comment by RedJue** [comment]
> 多了更多了，加这个功能多了 400B ..

要拆组件出来好像也确实避免不了...😅

File: components/modal/shared.tsx [hunk 15]
```diff
@@ -42,23 +45,48 @@ export const Footer: React.FC<
...
+        {typeof footer === 'function'
+          ? footer(footerOriginNode, {
+              ConfirmBtn: NormalOkBtn,
+              CancelBtn: NormalCancelBtn,
+            })
+          : footerOriginNode}
+      </ModalContextProvider>
     </DisabledContextProvider>
+  ) : (
+    footer
```
Comment (MadCcc) [inline components/modal/shared.tsx:90]: 这个是不是也应该放在 provider 里？

Comment (RedJue) [reply components/modal/shared.tsx:90]: 应该不用吧，这个footer应该是用户自己定义的节点，不应该受 provider 的影响

File: components/modal/shared.tsx [hunk 16]
```diff
@@ -42,23 +45,48 @@ export const Footer: React.FC<
...
-        loading={confirmLoading}
-        onClick={onOk}
-        {...okButtonProps}
-      >
-        {okText || locale?.okText}
-      </Button>
+      <ModalContextProvider value={btnCtxValueMemo}>
+        {typeof footer === 'function'
+          ? footer(footerOriginNode, {
+              ConfirmBtn: NormalOkBtn,
```
Comment (MadCcc) [inline components/modal/shared.tsx:83]:
```suggestion
              OkBtn: NormalOkBtn,
```

Comment (MadCcc) [reply components/modal/shared.tsx:83]: 命名统一一下？Modal 应该都叫 `ok`

Comment (RedJue) [reply components/modal/shared.tsx:83]: 嗯嗯 之前是根据 `issue` 留的 `API` 来写的 我改一下吧

**Review by MadCcc — APPROVED** [review]

**Pull request merged**